### PR TITLE
Fix decoding surrogates on error  #35073

### PR DESCRIPTION
**Pull request by buyaa-n**

### CODE REVIEW

File: src/System.Private.Xml/tests/XmlReader/Tests/System.Xml.RW.XmlReader.Tests.csproj [hunk 1]
```diff
@@ -7,6 +7,7 @@
     <Compile Include="AsyncReaderLateInitTests.cs" />
     <Compile Include="DisposeTests.cs" />
     <Compile Include="BaseUriTests.cs" />
+	<Compile Include="ReaderEncodingTests.cs" />
```
Comment (krwq) [inline src/System.Private.Xml/tests/XmlReader/Tests/System.Xml.RW.XmlReader.Tests.csproj]: fix indentation

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+using Xunit;
+
+namespace System.Xml.Tests
+{
+    public class ReaderEncodingTests
+    {
+        [Fact]
+        public static void ReadWithSurrogateCharAndInvalidChar()
+        {
+            var bytes = new byte[] { 60, 0, 0, 0, 0, 34, 1, 0, 62, 100, 60, 47, 97, 62, 10 };
```
Comment (krwq) [inline src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs:25]: if there is some reasonable textual representation for these I would prefer if you used that. I.e.:
`Encoding.Unicode.GetBytes("<?xml\u0123")`

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            var reader = XmlReader.Create(new MemoryStream(bytes));
+
+            XmlException ex = Assert.Throws<XmlException>(() => reader.Read());
+            Assert.Equal("Invalid character in the given encoding. Line 1, position 4.", ex.Message);
```
Comment (krwq) [inline src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs]: this won't work when using non en-US locale (although I think we currently don't support it). I'd just check if contains " 1" and " 4" perhaps or not check it at all

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            Assert.Equal(4, ex.LinePosition);
+        }
+
+        [Fact]
+        public static void ReadWithNormalCharAndInvalidChar()
+        {
+            var bytes = new byte[] { 60, 0, 0, 0, 65, 0, 0, 0,62, 100, 60, 47, 97, 62, 10 };
```
Comment (krwq) [inline src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs]: `0, 62` (space missing)

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+            XmlException ex = Assert.Throws<XmlException>(() => reader.Read());
+            Assert.Equal("Invalid character in the given encoding. Line 1, position 3.", ex.Message);
+            Assert.Equal(3, ex.LinePosition);
+        }
+
+        [Fact]
+        public static void ReadWithSurrogateCharAndInvalidChar_ValidXmlStructure()
+        {
+            var bytes = new byte[] { 60, 0, 0, 0, 97, 0, 0, 0, 62, 0, 0, 0, 0, 34, 1, 0, 62, 100, 60, 47, 60, 0, 0, 0, 47, 0, 0, 0, 97, 0, 0, 0, 62, 0, 0, 0 };
```
Comment (krwq) [inline src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs:49]: - all of these tests start with `60, 0, 0, 0`  should you check for invalid characters further in the buffer as well?
- I'm missing test where you try to make invalid character on the threshold of the buffer size
- positive cases proving you didn't break existing correct surrogate chars scenarios

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs:25]: Most of them doesn't have reasonable textual representation, actually the error happening while trying to convert these bytes to character

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs (hunk 3 above)
Comment (buyaa-n) [reply src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs]: Yeah I wasn't so confident with asserting error messages. Must be not consistent all time, actually we don't need these assertions, point of the test is there is no Error/Exception while decoding the bytes but expect XmlException while parsing to xml structure, so might be better to remove these assertions.

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs (hunk 5 above)
Comment (buyaa-n) [reply src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs:49]: > all of these tests start with 60, 0, 0, 0  should you check for invalid characters further in the buffer as well?

Didn't understood what you mean by "further in the buffer"

> I'm missing test where you try to make invalid character on the threshold of the buffer size

"on the threshold of the buffer size" is this mean at the end or beginning of byte array?

> positive cases proving you didn't break existing correct surrogate chars scenarios

There is one positive case https://github.com/buyaa-n/corefx/pull/1/files#diff-47e5e9b8251940e29f847159c2fd2fb4R55

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs (hunk 3 above)
Comment (krwq) [reply src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs]: ok, in such cases please also add test cases where invalid surrogates can be found somewhere in between of valid xml

Comment (buyaa-n) [reply src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs]: have that scenario here: https://github.com/buyaa-n/corefx/pull/1/files#diff-47e5e9b8251940e29f847159c2fd2fb4R55

File: src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.Private.Xml/tests/XmlReader/Tests/ReaderEncodingTests.cs:25]: added comment about which char is surrogate, which is invalid, which is normal char

**Review by krwq — APPROVED** [review]

**Pull request closed without merging**